README.md
## Rclone Remount v1.8
---

Remount cloud storage locally during boot via rclone & fusermount directly on your Android powered smart device. 

Virtually limitless storage expansion with support for dozens of cloud providers including Dropbox, GDrive, OneDrive, SFTP & many more. Extremely useful for devices without physical storage expansion capabilities. Also great for streaming large media files without need for full caching.  Binaries compiled using Termux. 

We are constantly striving to improve this project & make it the best. If you experience any issues or have suggestions please file them  [HERE](https://github.com/Magisk-Modules-Repo/com.piyushgarg.rclone/issues). Contributions to this project are welcomed. 

## Features

- Support for arm, arm64, ~~& x86~~

- Huge list of supported cloud storage providers

- Apps with ability to specify paths can access `/mnt/cloud/`

- Most file explorers work just fine ([issue #9](https://github.com/Magisk-Modules-Repo/com.piyushgarg.rclone/issues/9))

- Mount points use names of remote(s) in rclone.conf

- Specify custom rclone params for each remote via `/sdcard/.rclone/.REMOTE.param`

- Access remotes via [http://127.0.0.1:38762](http://127.0.0.1:38762)

- Access remotes via [ftp://127.0.0.1:38763](ftp://127.0.0.1:38763)

- Mount bind to `/sdcard/` (see [ issue #5](https://github.com/Magisk-Modules-Repo/com.piyushgarg.rclone/issues/5))

- Support for Work Profiles 

---
## Configuration (pre-installing)

1. Copy your `rclone.conf` file to `/sdcard/.rclone/rclone.conf` (not required)

2. Add custom params at `/sdcard/.rclone/.*.param` (if needed)

3. Install the module via Magisk Manager

4. Run `rclone config` via term if additional setup required 

4. All your rclone mount points will show up under `/mnt/cloud/` & `/storage/cloud/` or `/sdcard/cloud/`

For more detailed configuration of rclone please refer to [official documentation](https://rclone.org)

---
## Custom Params
Custom params have been created as a means for users to adjust this modules default parameters which are set for all remotes inside your rclone.conf. 

Specification of rclone parameters on a per remote basis can be created inside hidden files ending with the `.param` extension

      /sdcard/.rclone/.*.param

   Where `*` is replaced with the name of remote

- Custom parameters, their default values & rclone params they represent in `( )`

        LOGFILE=/sdcard/.rclone/rclone.log  ( --log-file )

        LOGLEVEL=NOTICE  ( --log-level )

        CACHEMODE=off  ( --vfs-cache-mode )

        CHUNKSIZE=1M  ( --cache-chunk-size )

        CHUNKTOTAL=1G  ( --cache-chunk-total-size )

        READCHUNKSIZE=1M  ( --vfs-read-chunk-size )

        CACHEWORKERS=1 ( --cache-workers )

        CACHEINFOAGE=1h0m0s  ( --cache-info-age )

        DIRCACHETIME=30m0s  ( --dir-cache-time )

        ATTRTIMEOUT=30s  ( --attr-timeout)

        BUFFERSIZE=0  ( --buffer-size )

        READAHEAD=128k  ( --max-read-ahead )

        M_UID=0  ( --uid )

        M_GID=1015  ( --gid )

        DIRPERMS=0775  ( --dir-perms )

        FILEPERMS=0644  ( --file-perms )

        UMASK=002  ( --umask )

        BINDSD=0  ( default binds remote to /sdcard/Cloud/* )

        SDBINDPOINT=  ( relative to /storage/emulated/0)

        SDSYNCDIRS= (relative to /storage/emulated/0)

        ADD_PARAMS=0

        REPLACE_PARAMS=0

        PROFILE=0

   **NOTE:** _The above are defaults for all remotes without `.*.param` files containing opposing values. 

- Custom remote params example #1

   _The following configuration will disable caching for remote `[Movies]`, bind to `/sdcard/Movies` & add the `-fast-list`/`--allow-non-empty` flags to it's mounting command._

         /sdcard/.rclone/.Movies.param

         1| CACHEMODE=off
         2| BINDSD=1
         3| SDBINDPOINT=Movies
         4| ADD_PARAMS=--fast-list --allow-non-empty
         5| 

    **NOTE:** _There is no need to specify values you do not wish to change. Ensure a line break/carriage return exist after each specified param or they will not be parsed. For more information see [issue #2](https://github.com/Magisk-Modules-Repo/com.piyushgarg.rclone/issues/2)_

---
## Custom Globals

Specification of global rclone parameters can be created in

      /sdcard/.rclone/.global.param

- Global Specific Parameters

        NETCHK=1

        NETCHK_ADDR=google.com

        NETCHK_IF=

        HTTP=1

        HTTP_ADDR=127.0.0.1:38762

        FTP=1

        FTP_ADDR=127.0.0.1:38763

- Excluded Parameters

        SDBINDPOINT=
        SDSYNCDIRS=

- Custom globals params example #1

  _The following configuration will enable minimal caching for all remotes, bind to `/sdcard/Cloud/*`, disable HTTP/FTP & add the `--fast-list`/`--allow-non-empty` flags to their mounting command(s)._

         /sdcard/.rclone/.global.param

         1| CACHEMODE=minimal
         2| BINDSD=1
         3| ADD_PARAMS=--fast-list --allow-non-empty
         4| HTTP=0
         5| FTP=0
         6| 

   **NOTE:** _Global parameters effect all remotes without `.*.parm` files containing opposing values. Some parameters are specific to globals while others have been excluded._

---
## Replacing & Adding Params 

In order for users to  appropriately utilize  `ADD_PARAMS=` or `REPLACE_PARAMS=` they will need a little background on the parameters that are set by default. 

- Currently specified params shown here ↓

  (directly from service.sh):

  `RCLONE_PARAMS=" --log-file ${LOGFILE} --log-level ${LOGLEVEL} --vfs-cache-mode ${CACHEMODE} --cache-dir ${CACHE} --cache-chunk-path ${CACHE_BACKEND} --cache-db-path ${CACHE_BACKEND} --cache-tmp-upload-path ${CACHE} --vfs-read-chunk-size ${READCHUNKSIZE} --vfs-cache-max-size ${CACHEMAXSIZE} --cache-chunk-size ${CHUNKSIZE} --cache-chunk-total-size ${CHUNKTOTAL} --cache-workers ${CACHEWORKERS} --cache-info-age ${CACHEINFOAGE} --dir-cache-time ${DIRCACHETIME} --attr-timeout ${ATTRTIMEOUT} --cache-chunk-no-memory --use-mmap --buffer-size ${BUFFERSIZE} --max-read-ahead ${READAHEAD} --no-modtime --no-checksum --uid ${M_UID} --gid ${M_GID} --allow-other --dir-perms ${DIRPERMS} --file-perms ${FILEPERMS} --umask ${UMASK} ${READONLY} ${ADD_PARAMS} "`

                                ^

    **NOTE:** _When using the `ADD_PARAMS=` it will append any additonal params you wish to specify at the point of `${ADD_PARAMS}` (above) in a fill in the blank manner._

- The script then takes `RCLONE_PARAMS=` and fills in blank at `${RCLONE_PARAMS}`

  `rclone mount ${remote}: ${CLOUDROOTMOUNTPOINT}/${remote} --config ${CONFIGFILE} ${RCLONE_PARAMS} --daemon &`

  **NOTE:** _Everything before and after `${RCLONE_PARAMS}` cannot not be replaced even with `REPLACE_PARAMS=` specified._

- When using `REPLACE_PARAMS=` `RCLONE_PARAMS=` becomes `RCLONE_PARAMS=" ${REPLACE_PARAMS} "`

---
## Work Profiles & Users

As of `v1.8` support for isolating & binding to work profiles or additional users has been included which may provide for some interesting use cases. 

When adding work profiles through sandboxing apps such as [Island](https://play.google.com/store/apps/details?id=com.oasisfeng.island) or [Shelter](https://play.google.com/store/apps/details?id=net.typeblog.shelter) it will create a virtual SD for your sandboxed apps. This virtual SD can now be used with rclone remount. 

- Work profile example #1 (Cloud Camera w/ Shelter)

        open Shelter > find camera > clone to work profile

        /sdcard/.rclone/.cloud-DCIM.param

        1| BINDSD=1
        2| SDBINDPOINT=DCIM
        3| PROFILE=10
        4| ISOLATE=1
        5| CACHEMODE=writes
        6| 

   **NOTE:** _Virtual SDs for work profiles & or additional users start at `/storage/emulated/`**10**. Additional profiles increase the ending directory integer (e.g. `/storage/emulated/`**11**). This integer is used with `PROFILE=`_
---
## SD Sync & Remotes

- SD sync example # 1

         /sdcard/.rclone/.Backup.param

         1| SDSYNCDIRS=DCIM/Camera:Photos:My Projects
         2| CACHEMODE=writes
         3| CACHEINFOAGE=11s
         4| DIRCACHETIME=10s
         5| ATTRTIMEOUT=10s
         6| 

  **NOTE:** _`SDSYNCDIRS=` paths are relative to /storage/emulated/`PROFILE=0`. Paths are to be separated using a `: `. This variable is should be whitespace friendly._
---
## Known Issues

- VLC  takes a long time to load media as it opens file in write mode when using it's internal browser. 

   a. Create remote type alias for media dirs in rclone.conf and 
specify `CACHEMODE=off` in `/sdcard/.rclone/.ALIASNAME.param`

- Encrypted devices can not mount until unlock

- Encrypted `rclone.conf` causes reboots

- High cpu/mem in some apps with storage perms ([issue #9](https://github.com/Magisk-Modules-Repo/com.piyushgarg.rclone/issues/9))
- The `fusermount` bin may not be compatible on all devices (see  [thread](https://www.google.com/amp/s/forum.xda-developers.com/android/development/fusermount-android-rclone-mount-t3866652/amp/))

---
## Disclaimer

Neither the author nor developer's will be held responsible for any damage/data loss that may occur during use of this module. While we have done our best to make sure no harm will come about, no guarantees can be made. Keep in mind the binaries included in this project were originally intended to be ran on PCs which may cause unforseen issues. Always check this document before updating to new releases as significant changes may occur. 

---
## Credits

- rclone devs
- pmj_pedro[@xda](https://forum.xda-developers.com/showpost.php?p=78147335&postcount=1)
- agnostic-apollo[@xda](https://forum.xda-developers.com/showpost.php?p=79929083&postcount=12)
- rclone binaries from [rclone.org](https://rclone.org/downloads)
- fusermount binaries from  [xda-devs](https://forum.xda-developers.com/android/development/fusermount-android-rclone-mount-t3866652)
- improvements geofferey@github

---
## Changelog

## v1.13
* Add arm/arm64 1.53 binaries downloaded from https://beta.rclone.org/v1.53.0/testbuilds/rclone-android-16-arm.gz

## v1.12
* Fixed restart problems.

## v1.11
* Add arm/arm64 1.52 bins downloaded from https://beta.rclone.org/v1.52.0/testbuilds/rclone-android-16-arm.gz
* Fixed service.sh paths

## v1.10
* fixed fusermount wrapper

## v1.9
* Add arm/arm64 1.51 bins downloaded from https://beta.rclone.org/
* Commented fusermount wrapper

### v1.8
* Support for Work Profiles `PROFILE=`
* Isolate to Work Profiles `ISOLATE=1`
* Support syncing from SD to remote 

### v1.7
* Add ability to disable HTTP/FTP
* Link rest of default params to custom vars
* Exclude some custom params from globals
* Make some globals exclusive 
* Change `BINDPOINT=` to `SDBINDPOINT=`
* Fix bug with custom params
* Set `PATH=` to change priority of used bins

### v1.6
* Simplify custom global parameters
* Fix & improve binding to SD
* Specify additional  rclone ops with `ADD_PARAMS=`
* Replace `rclone mount` ops via `REPLACE_PARAMS=`

### v1.5
* Replace arm/arm64  `rclone` 1.48 bins built with Termux
* Replace arm/arm64 `fusermount` built with Termux
* Add arm/arm64 `libandroid-support.so` from Termux
* Support for mounting to SD
* Squash missing rclone.conf install bug
* Tune default parameters
* Include a wrap for `rclone config`
* Include `fusermount-wrapper.sh`
* General Improvements

### v1.4
* Add ability to disable a remote 
* Add a wrapper script for rclone
* Access remotes via http & ftp
* Use without rebooting device
* Add wrapper cmds to `rclone help`
* Make remount possible via `su -M -c`

### v1.3
* Move user rclone.conf & related to `/sdcard/.rclone/`
* Control global `--vfs-cache-mode` via simple files placed in `/sdcard/.rclone/`
* Specify custom params for individual remotes via `/sdcard/.rclone/.REMOTENAME.params`

### v1.2
* Change install process
* Changes for full systemless
* Improve mount reliability
* Symlink mountpoint to `/storage/`

### v1.1
* Initial release
* rclone mount

[![HitCount](http://hits.dwyl.io/Magisk-Modules-Repo/compiyushgargrclone.svg)](http://hits.dwyl.io/Magisk-Modules-Repo/compiyushgargrclone)


changelog.md
# Changelog
## v1.9
* Add arm/arm64 1.51 bins downloaded from https://beta.rclone.org/
* Commented fusermount wrapper

## v1.5
* Add arm/arm64 1.48 bins compiled using Termux
* Add static arm64 `fusermount`
* Support for mounting to SD
* Squash missing rclone.conf install bug
* Tune default parameters
* Include a wrap for `rclone config`

## v1.4
* Add ability to disable a remote
* Add a wrapper script for rclone
* Make remount possible via `su --mount-master`

## v1.3
* Move user rclone.conf & related to /sdcard/.rclone/
* Control global --vfs-cache-mode via simple files placed in /sdcard/.rclone/
* Specify custom params for individual remotes via /sdcard/.rclone/.REMOTENAME.params

## v1.2
* Change install process
* Changes for full systemless
* Improve mount reliability
* Link to /storage/

## v1.1
* Initial release
* rclone mount


common/service.sh
#!/system/bin/sh
# Do NOT assume where your module will be located.
# ALWAYS use $MODDIR if you need to know where this script
# and module is placed.
# This will make sure your module will work
# if Magisk changes it's mount point in the future

PATH=/system/bin:/sbin:/sbin/.core/busybox:/system/xbin

MODDIR="$(dirname "$(readlink -f "$0")")"

IMGDIR=/sbin/.core/img
UPDDIR=/data/adb/modules_update
id=com.piyushgarg.rclone

## resolve links - $0 may be a link to module home
PRG="$0"

# need this for relative symlinks
while [ -h "$PRG" ] ; do
  ls=`ls -ld "$PRG"`
  link=`expr "$ls" : '.*-> \(.*\)$'`
  if expr "$link" : '/.*' > /dev/null; then
    PRG="$link"
  else
    PRG="`dirname "$PRG"`/$link"
  fi
done

saveddir=`pwd`
MODDIR2=`dirname "$PRG"`
# make it fully qualified
MODDIR2=`cd "$MODIR2" && pwd`
cd "$saveddir"
#echo 3 $MODDIR2

if [ -e ${UPDDIR}/${id}/rclone ]; then
    HOME=${UPDDIR}/${id}

elif [ -e ${IMGDIR}/${id}/rclone ]; then
    HOME=${IMGDIR}/${id}

elif [ -e ${MODDIR2}/${id}/rclone ]; then
    HOME=${MODDIR2}/${id}

else
    HOME=${MODDIR}
fi

echo $HOME

ln -sf ${HOME}/rclone /sbin/rclone
ln -sf ${HOME}/fusermount /sbin/fusermount
ln -sf ${HOME}/rclone-wrapper.sh /sbin/rclonew

#MODULE VARS
SYSBIN=/system/bin
CLOUDROOTMOUNTPOINT=/mnt/cloud
USER_CONFDIR=/sdcard/.rclone
USER_CONF=${USER_CONFDIR}/rclone.conf
PROFILE=0
DATA_MEDIA=/data/media
RUNTIME_R=/mnt/runtime/read
RUNTIME_W=/mnt/runtime/write
RUNTIME_D=/mnt/runtime/default
BINDPOINT_R=${RUNTIME_R}/emulated/${PROFILE}/Cloud
BINDPOINT_W=${RUNTIME_W}/emulated/${PROFILE}/Cloud
BINDPOINT_D=${RUNTIME_D}/emulated/${PROFILE}/Cloud
SD_BINDPOINT=${BINDPOINT_D}
DISABLE=0
NETCHK=1
NETCHK_ADDR=google.com

#RCLONE PARAMETERS
CONFIGFILE=${HOME}/.config/rclone/rclone.conf
LOGFILE=${USER_CONFDIR}/rclone.log
LOGLEVEL=NOTICE
CACHE=/data/rclone/cache
CACHE_BACKEND=/data/rclone/cache/cache-backend
CACHEMODE=off
READCHUNKSIZE=1M
CACHEMAXSIZE=1G
CHUNKSIZE=1M
CHUNKTOTAL=1G
CACHEWORKERS=1
CACHEINFOAGE=1h0m0s
DIRCACHETIME=30m0s
ATTRTIMEOUT=30s
BUFFERSIZE=0
READAHEAD=128k
M_UID=0
M_GID=1015
DIRPERMS=0775
FILEPERMS=0644
UMASK=002
BINDSD=0
SYNC_WIFI=1
SYNC_CHARGE=0
SYNC_BATTLVL=0
HTTP=0
HTTP_ADDR=127.0.0.1:38762
FTP=0
FTP_ADDR=127.0.0.1:38763
SFTP=0
SFTP_ADDR=127.0.0.1:38722
SFTP_USER=
SFTP_PASS=


if [[ -z ${INTERACTIVE} ]]; then

    INTERACTIVE=0

fi

if [[ ! -d ${HOME}/.config/rclone ]]; then

    mkdir -p ${HOME}/.config/rclone

fi

custom_params () {

    if [[ ${remote} = global ]]; then

        PARAMS="DISABLE LOGFILE LOGLEVEL CACHEMODE CHUNKSIZE CHUNKTOTAL CACHEWORKERS CACHEINFOAGE DIRCACHETIME ATTRTIMEOUT BUFFERSIZE READAHEAD M_UID M_GID DIRPERMS FILEPERMS READONLY BINDSD ADD_PARAMS REPLACE_PARAMS NETCHK NETCHK_IF NETCHK_ADDR HTTP FTP HTTP_ADDR FTP_ADDR SFTP SFTP_ADDR SFTP_USER SFTP_PASS PROFILE ISOLATE"

    else

        PARAMS="DISABLE LOGFILE LOGLEVEL CACHEMODE CHUNKSIZE CHUNKTOTAL CACHEWORKERS CACHEINFOAGE DIRCACHETIME ATTRTIMEOUT BUFFERSIZE READAHEAD M_UID M_GID DIRPERMS FILEPERMS READONLY BINDSD SDBINDPOINT ADD_PARAMS REPLACE_PARAMS PROFILE ISOLATE SDSYNCDIRS SYNC_WIFI SYNC_BATTLVL SYNC_CHARGE SUBPATH"
    fi

    BAD_SYNTAX="(^\s*#|^\s*$|^\s*[a-z_][^[:space:]]*=[^;&\(\`]*$)"

    if [[ -e ${USER_CONFDIR}/.${remote}.param ]]; then

        echo "Found .${remote}.param"

        if ! [[ $(egrep -q -iv "${BAD_SYNTAX}" ${USER_CONFDIR}/.${remote}.param) ]]; then

            echo "loading .${remote}.param"

            # FIX: Unnecessary and very inefficient double loop
            for PARAM in ${PARAMS[@]}; do

                while read -r VAR; do

                    if [[ "$(echo "${VAR}" |grep "$PARAM=")" ]]; then
                        echo "Importing ${VAR}"
                        VALUE="$(echo ${VAR} |cut -d '=' -f2)"

                        VALUE=\"${VALUE}\"

                        # Unnecessary echo in a subshell execution below? Why not just: 
                        # eval "${PARAM}""=""${VALUE}"
                        eval $(echo "${PARAM}""=""${VALUE}")

                    fi

                done < ${USER_CONFDIR}/.${remote}.param

            done

        else

            echo ".${remote}.param contains bad syntax"

        fi

    fi

}

global_params () {

    remote=global
    custom_params
    unset remote
    echo

}

net_chk() {

    if [ -z ${NETCHK_IF} ]; then

        NETCHK_IF=" "

    else

        NETCHK_IF=" -I ${NETCHK_IF} "
    
    fi

    ping ${NETCHK_IF} -c 5 ${NETCHK_ADDR}

}

sd_unbind () {

    if [[ -z ${SDBINDPOINT} ]]; then

        UNBINDPOINT=${BINDPOINT_D}/${remote}

        su -M -c umount -lf ${UNBINDPOINT} >> /dev/null 2>&1

        UNBINDPOINT=${BINDPOINT_R}/${remote}

        su -M -c umount -lf ${UNBINDPOINT} >> /dev/null 2>&1

        UNBINDPOINT=${BINDPOINT_W}/${remote}

        su -M -c umount -lf ${UNBINDPOINT} >> /dev/null 2>&1

    else 

        USER_BINDPOINT=${SDBINDPOINT}

        UNBINDPOINT=${RUNTIME_D}/emulated/${PROFILE}/${USER_BINDPOINT}

        su -M -c umount -lf ${UNBINDPOINT} >> /dev/null 2>&1

        UNBINDPOINT=${RUNTIME_R}/emulated/${PROFILE}/${USER_BINDPOINT}

        su -M -c umount -lf ${UNBINDPOINT} >> /dev/null 2>&1

        UNBINDPOINT=${RUNTIME_W}/emulated/${PROFILE}/${USER_BINDPOINT}

        su -M -c umount -lf ${UNBINDPOINT} >> /dev/null 2>&1

        fi

}

sd_binder () {

    if [[ -d ${RUNTIME_D} ]] && [[ ${BINDSD} = 1 ]]; then

        if [[ -z ${SDBINDPOINT} ]]; then 

            mkdir -p ${DATA_MEDIA}/${PROFILE}/Cloud/${remote}
            chown media_rw:media_rw ${DATA_MEDIA}/${PROFILE}/Cloud/$remote

            BINDPOINT=${BINDPOINT_D}/${remote}

            su -M -c mount --bind ${CLOUDROOTMOUNTPOINT}/${remote} ${BINDPOINT} >> /dev/null 2>&1

            BINDPOINT=${BINDPOINT_R}/${remote}

            if ! mount |grep -q ${BINDPOINT}; then

                su -M -c mount --bind ${CLOUDROOTMOUNTPOINT}/${remote} ${BINDPOINT} >> /dev/null 2>&1

            fi

            BINDPOINT=${BINDPOINT_W}/${remote}

            if ! mount |grep -q ${BINDPOINT}; then

            su -M -c mount --bind ${CLOUDROOTMOUNTPOINT}/${remote} ${BINDPOINT} >> /dev/null 2>&1

            fi
            
            echo "[$remote] available at: -> [/sdcard/Cloud/${remote}]"

        else 

            mkdir ${DATA_MEDIA}/${PROFILE}/${SDBINDPOINT} >> /dev/null 2>&1
            chown media_rw:media_rw ${DATA_MEDIA}/${PROFILE}/${SDBINDPOINT}

            USER_BINDPOINT=${SDBINDPOINT}
            BINDPOINT=${RUNTIME_D}/emulated/${PROFILE}/${USER_BINDPOINT}

            su -M -c mount --bind ${CLOUDROOTMOUNTPOINT}/${remote} ${BINDPOINT} >> /dev/null 2>&1

            BINDPOINT=${RUNTIME_R}/emulated/${PROFILE}/${USER_BINDPOINT}

            if ! mount |grep -q ${BINDPOINT}; then

                su -M -c mount --bind ${CLOUDROOTMOUNTPOINT}/${remote} ${BINDPOINT} >> /dev/null 2>&1

            fi

            BINDPOINT=${RUNTIME_W}/emulated/${PROFILE}/${USER_BINDPOINT}

            if ! mount |grep -q ${BINDPOINT}; then

                su -M -c mount --bind ${CLOUDROOTMOUNTPOINT}/${remote} ${BINDPOINT} >> /dev/null 2>&1

            fi

            echo "[$remote] available at: -> [/storage/emulated/${PROFILE}/${SDBINDPOINT}]"
            
            unset BINDPOINT

        fi

    fi
}

syncd_service () {

    if [[ ! -z ${SDSYNCDIRS} ]]; then

        export PIDFILE=${HOME}/.tmp/${remote}-syncd.pids

        kill -9 $(cat ${PIDFILE}) >> /dev/null 2>&1
        rm ${PIDFILE} >> /dev/null 2>&1

        if [[ ! -d ${HOME}/.tmp ]]; then

            mkdir -p ${HOME}/.tmp

        fi

    export PATH
    export CLOUDROOTMOUNTPOINT
    export PROFILE
    export HOME
    export remote
    export SYNCWIFI
    export SYNC_BATTLVL
    export SYNC_CHARGE
    export NETCHK_ADDR

        IFS=$':'

        for SYNCDIR in ${SDSYNCDIRS[@]}; do

            export SYNCDIR

            ${HOME}/syncd.sh >> /dev/null 2>&1 &

        done

    fi

    unset IFS
    unset SDSYNCDIRS
    unset SYNCDIR
    SYNC_BATTLVL=0
    SYNC_WIFI=1
    SYNC_CHARGE=0

}

reset_params () {

    unset IFS
    unset SDBINDPOINT
    unset BINDSD
    unset ISOLATE
    unset RCLONE_PARAMS
    unset REPLACE_PARAMS
    unset ADD_PARAMS
    unset SYNCDIR
    unset SDSYNCDIRS
    unset PIDFILE
    unset SUBPATH
    LOGFILE=${USER_CONFDIR}/rclone.log
    LOGLEVEL=NOTICE
    CACHEMODE=off
    READCHUNKSIZE=1M
    CACHEMAXSIZE=1G
    CHUNKSIZE=1M
    CHUNKTOTAL=1G
    CACHEWORKERS=1
    CACHEINFOAGE=1h0m0s
    DIRCACHETIME=30m0s
    ATTRTIMEOUT=30s
    BUFFERSIZE=0
    READAHEAD=128k
    M_UID=0
    M_GID=1015
    DIRPERMS=0775
    FILEPERMS=0644
    UMASK=002
    PROFILE=0
    SYNC_WIFI=1
    SYNC_BATTLVL=0
    SYNC_CHARGE=0

}

rclone_mount () {

    if [[ ${READONLY} = 1 ]]; then

        READONLY=" --read-only "

    else

        READONLY=" "

    fi

    if [[ ${ADD_PARAMS} = 0 ]]; then

        unset ADD_PARAMS

    fi

    if [[ ${REPLACE_PARAMS} = 0 ]]; then

        unset REPLACE_PARAMS

    fi

    if [[ -z ${REPLACE_PARAMS} ]]; then

        RCLONE_PARAMS=" --log-file ${LOGFILE} --log-level ${LOGLEVEL} --vfs-cache-mode ${CACHEMODE} --cache-dir ${CACHE} --cache-chunk-path ${CACHE_BACKEND} --cache-db-path ${CACHE_BACKEND} --cache-tmp-upload-path ${CACHE} --vfs-read-chunk-size ${READCHUNKSIZE} --vfs-cache-max-size ${CACHEMAXSIZE} --cache-chunk-size ${CHUNKSIZE} --cache-chunk-total-size ${CHUNKTOTAL} --cache-workers ${CACHEWORKERS} --cache-info-age ${CACHEINFOAGE} --dir-cache-time ${DIRCACHETIME} --attr-timeout ${ATTRTIMEOUT} --cache-chunk-no-memory --use-mmap --buffer-size ${BUFFERSIZE} --max-read-ahead ${READAHEAD} --no-modtime --no-checksum --uid ${M_UID} --gid ${M_GID} --allow-other --dir-perms ${DIRPERMS} --file-perms ${FILEPERMS} --umask ${UMASK} ${READONLY} ${ADD_PARAMS} "

    elif [[ ! -z ${REPLACE_PARAMS} ]]; then

        RCLONE_PARAMS=" ${REPLACE_PARAMS} "

    fi

    if [[ -z ${ADD_PARAMS} ]]; then

        ADD_PARAMS=" "

    elif [[ ! -z ${ADD_PARAMS} ]]; then

        ADD_PARAMS=" ${ADD_PARAMS} "

    fi

    echo "[${remote}] available at: -> [${CLOUDROOTMOUNTPOINT}/${remote}]"

    mkdir -p ${CLOUDROOTMOUNTPOINT}/${remote}

su -M -p -c nice -n 19 ionice -c 2 -n 7 ${HOME}/rclone mount ${remote}:${SUBPATH} ${CLOUDROOTMOUNTPOINT}/${remote} --config ${CONFIGFILE} ${RCLONE_PARAMS} --daemon >> /dev/null 2>&1 &

}

COUNT=0

if [[ ${INTERACTIVE} = 0 ]]; then

    until [[ $(getprop sys.boot_completed) = 1 ]] && [[ $(getprop dev.bootcomplete) = 1 ]] && [[ $(getprop init.svc.bootanim) = stopped ]] && [[ -e ${USER_CONF} ]] || [[ ${COUNT} -eq 240 ]]; do

        sleep 5
        ((++COUNT))

    done

fi

DECRYPT_CHK () {

    su -M -c ls sdcard |grep -q -w "Android"

}

if [[ ${COUNT} -eq 240 ]] || [[ ! -d /sdcard/Android ]]; then

    exit 0

fi

if [[ ! -d ${USER_CONFDIR} ]]; then 

    mkdir ${USER_CONFDIR}

fi

if [[ -e ${USER_CONFDIR}/.disable ]] && [[ ${INTERACTIVE} = 0 ]]; then 

    exit 0

fi

if [[ ${INTERACTIVE} = 0 ]]; then

    export INTERACTIVE=1

    ${HOME}/service.sh

    exit

fi

global_params

if [[ ! -d ${CLOUDROOTMOUNTPOINT} ]]; then

    mkdir -p ${CLOUDROOTMOUNTPOINT}
    chown root:sdcard_rw ${CLOUDROOTMOUNTPOINT}
    touch ${CLOUDROOTMOUNTPOINT}/.nomedia
    chown root:sdcard_rw ${CLOUDROOTMOUNTPOINT}/.nomedia
    chmod 0775 ${CLOUDROOTMOUNTPOINT}/.nomedia

fi

if [[ ! -e ${CLOUDROOTMOUNTPOINT}/.nomedia ]]; then

    touch ${CLOUDROOTMOUNTPOINT}/.nomedia
    chown root:sdcard_rw ${CLOUDROOTMOUNTPOINT}/.nomedia
    chmod 0644 ${CLOUDROOTMOUNTPOINT}/.nomedia

fi

if [[ ! -d ${CACHE} ]]; then

    mkdir -p ${CACHE}
    chown root:sdcard_rw ${CACHE}
    chmod 0775 ${CACHE}

fi

if [[ -d ${CACHE} ]]; then

    chown root:sdcard_rw ${CACHE}
    chmod 0775 ${CACHE}

fi

if [[ ! -d ${CACHE_BACKEND} ]]; then

    mkdir -p ${CACHE_BACKEND}

fi

if [[ -d ${CACHE_BACKEND} ]]; then

    chown root:sdcard_rw ${CACHE_BACKEND}
    chmod 0775 ${CACHE_BACKEND}

fi

if [[ ! -L ${RUNTIME_R}/cloud ]]; then

    ln -sf ${CLOUDROOTMOUNTPOINT} ${RUNTIME_R}/cloud

fi

if [[ ! -L ${RUNTIME_W}/cloud ]]; then

    ln -sf ${CLOUDROOTMOUNTPOINT} ${RUNTIME_W}/cloud

fi

if [[ ! -L ${RUNTIME_D}/cloud ]]; then

    ln -sf ${CLOUDROOTMOUNTPOINT} ${RUNTIME_D}/cloud

fi

if [[ -e ${USER_CONF} ]]; then

    cp ${USER_CONF} ${CONFIGFILE}
    chmod 0600 ${CONFIGFILE}

fi

if [[ ${NETCHK} = 1 ]]; then

    until net_chk || [[ ${COUNT} = 60 ]]; do 

        sleep 5
        ((++COUNT))

    done >> /dev/null 2>&1

fi

echo "Default CACHEMODE ${CACHEMODE}"

sleep 5

LD_LIBRARY_PATH=${HOME} ${HOME}/rclone listremotes --config ${CONFIGFILE}|cut -f1 -d: | 

    while read remote; do

        echo

        list_remote=${remote}
        CLOUDROOTMOUNTPOINT=/mnt/cloud
        PROFILE=0
        DISABLE=0
        READONLY=0

        global_params >> /dev/null 2>&1

        remote=${list_remote}

        custom_params

        if [[ ${ISOLATE} = 1 ]] && [[ ${PROFILE} -gt 0 ]] && [[ ${BINDSD} = 1 ]]; then

            CLOUDROOTMOUNTPOINT=/data/media/${PROFILE}/.cloud

        fi

        if [[ ${DISABLE} = 1 ]] || [[ -e ${USER_CONFDIR}/.${remote}.disable ]]; then

            echo "${remote} disabled by user"
            continue

        fi

        sd_unbind
        rclone_mount
        sd_binder
        syncd_service
        reset_params

    done

echo

if [[ ${HTTP} = 1 ]]; then

    if $(/sbin/rclone serve http ${CLOUDROOTMOUNTPOINT} --addr ${HTTP_ADDR} --no-checksum --no-modtime --read-only >> /dev/null 2>&1 &); then

        echo "Notice: /mnt/cloud served via HTTP at: http://${HTTP_ADDR}"

    fi

fi

if [[ ${FTP} = 1 ]]; then

    if $(/sbin/rclone serve ftp ${CLOUDROOTMOUNTPOINT} --addr ${FTP_ADDR} --no-checksum --no-modtime --read-only >> /dev/null 2>&1 &); then

        echo "Notice: /mnt/cloud served via FTP at: ftp://${FTP_ADDR}"

    fi
    
fi

if [[ ${SFTP} = 1 ]] && [[ ! -z ${SFTP_USER} ]] && [[ ! -z ${SFTP_PASS} ]]; then

    if $(/sbin/rclone serve sftp ${CLOUDROOTMOUNTPOINT} --addr ${SFTP_ADDR} --user ${SFTP_USER} --pass ${SFTP_PASS} --no-checksum --no-modtime --read-only >> /dev/null 2>&1 &); then

        echo "Notice: /mnt/cloud served via SFTP at: sftp://${SFTP_ADDR}"

    fi

fi

echo
echo "...done"

exit
